sensing/pointcloud_preprocessor/src/outlier_filter/ring_outlier_filter_nodelet.cpp

```diff
@@ -44,82 +44,116 @@ RingOutlierFilterComponent::RingOutlierFilterComponent(const rclcpp::NodeOptions
-        *reinterpret_cast<float *>(&input_ptr->data[next_idx + azimuth_offset]);
-      float azimuth_diff = next_pt_azimuth - current_pt_azimuth;
+
+      const float & current_azimuth =
+        *reinterpret_cast<const float *>(&input->data[current_data_idx + azimuth_offset]);
+      const float & next_azimuth =
+        *reinterpret_cast<const float *>(&input->data[next_data_idx + azimuth_offset]);
+      float azimuth_diff = next_azimuth - current_azimuth;
       azimuth_diff = azimuth_diff < 0.f ? azimuth_diff + 36000.f : azimuth_diff;
 
-      const auto current_pt_distance =
-        *reinterpret_cast<float *>(&input_ptr->data[current_idx + distance_offset]);
-      const auto next_pt_distance =
-        *reinterpret_cast<float *>(&input_ptr->data[next_idx + distance_offset]);
+      const float & current_distance =
+        *reinterpret_cast<const float *>(&input->data[current_data_idx + distance_offset]);
+      const float & next_distance =
+        *reinterpret_cast<const float *>(&input->data[next_data_idx + distance_offset]);
 
       if (
-        std::max(current_pt_distance, next_pt_distance) <
-          std::min(current_pt_distance, next_pt_distance) * distance_ratio_ &&
+        std::max(current_distance, next_distance) <
+          std::min(current_distance, next_distance) * distance_ratio_ &&
         azimuth_diff < 100.f) {
-        continue;
+        continue;  // Determined to be included in the same walk
       }
-      if (isCluster(input_ptr, tmp_indices)) {
-        for (const auto & tmp_idx : tmp_indices) {
-          output_modifier.push_back(
-            std::move(*reinterpret_cast<PointXYZI *>(&input_ptr->data[tmp_idx])));
+
+      if (isCluster(
+            input, indices[walk_first_idx], indices[walk_last_idx],
+            walk_last_idx - walk_first_idx + 1)) {
+        for (int i = walk_first_idx; i <= walk_last_idx; i++) {
+          auto output_ptr = reinterpret_cast<PointXYZI *>(&output.data[output_size]);
+          *output_ptr = *reinterpret_cast<const PointXYZI *>(&input->data[indices[i]]);
+          output_size += output.point_step;
         }
       }
-      tmp_indices.clear();
-    }
-    if (tmp_indices.empty()) {
-      continue;
+
+      walk_first_idx = idx + 1;
     }
-    if (isCluster(input_ptr, tmp_indices)) {
-      for (const auto & tmp_idx : tmp_indices) {
-        output_modifier.push_back(
-          std::move(*reinterpret_cast<PointXYZI *>(&input_ptr->data[tmp_idx])));
+
+    if (walk_first_idx > walk_last_idx) continue;
+
+    if (isCluster(
+          input, indices[walk_first_idx], indices[walk_last_idx],
+          walk_last_idx - walk_first_idx + 1)) {
+      for (int i = walk_first_idx; i <= walk_last_idx; i++) {
+        auto output_ptr = reinterpret_cast<PointXYZI *>(&output.data[output_size]);
+        *output_ptr = *reinterpret_cast<const PointXYZI *>(&input->data[indices[i]]);
+        output_size += output.point_step;
       }
     }
-    tmp_indices.clear();
   }
```

Comment:
Another approximation I have found with this implementation is that the ring end is not "connected" to its start.

For example, let's imagine this ring:
```
ring : xxxx..xxxxxxxxxxx..xxxx.xx
index: abcdefghijklmnopqrstuvwxyz
``` 
Where chains of `x` represent points from the same "walk", and `.` the noisy points ment to be filtered. The index alphabet is just for the explanation. So in the situation here, there are 3 rings:
1. from 'g' to 'q'
2. from 't' to 'w'
3. from 'y' to 'd' (wrapped around)

However in your implementation, we start at index 0 ('a') and the start/end of the ring are not considered connected. So you would find the rings:
1. from 'a' to 'd'
2. from 'g' to 'q'
3. from 't' to 'w'
The algorithm would consider the walk 'y' -> 'z', but as it is too short it would ignore the points.

With the default parameters `num_points_threshold_ = 4`, up to 3 points could be missing because of that. When using bigger `num_points_threshold_` value, the filter would sometimes produce some big blinspot in the data.

Reply:
Since this problem has been around for a while and has nothing to do with this Pull Request, can you please create a new Issue?